### DISCUSSION TITLE
How can I tell if a character doesn't exist in the font?

### DISCUSSION THREAD

**Discussion by freakypie** [protectwise/troika#173] [Q&A]
I'm writing an app that uses Latin, Cyrillic, and Chinese characters. If a character fails to load it loads a box, but I haven't identified a consistent way to determine if I'm getting the "Not found" box or a real character.

Is there a built in way to determine if a character failed to load?

**Comment by lojjic** [comment]
Interesting question. My initial reaction is that when you see a box like that it's almost always going to be the fallback glyph for an absent character -- it's possible for it to be some unicode symbol character but it seems like that would be so rare you could usually ignore it ...?

Otherwise, I've used https://fontdrop.info/ as a convenient way to inspect what's included in a given font file, you could try that.

If you're looking for a programmatic way to determine absent characters, that info isn't currently exposed but it wouldn't be a difficult addition to the TextRenderInfo struct.

Also maybe of interest: https://github.com/protectwise/troika/issues/13

**Reply by freakypie** [reply]
My users are international, so I can't be sure which characters they will use beforehand. I want to fallback to a more inclusive font if a character is not found. 

If you can point me to the right spot in the code, I might could do a pull request. I tried to find it before and I failed

**Reply by lojjic** [reply]
Ah, thanks for clarifying.

Hypothetically, if #13 was implemented, would that cover your use case? I feel like it would -- e.g. if your users entered text with cyrillic characters then it would detect that and load the additional font file that contains that character range, all seamless behind the scenes.

That feature is pretty high on my priority list; I may be tempted to bump it to the top of my list if that would meet your immediate needs...?

**Reply by freakypie** [reply]
Possible. Where my app can have user uploaded fonts, there might be an issue with them not knowing what unicode ranges their font would have, I have only the most minor understanding of unicode myself.

So #13 would help, but it still would be helpful to detect that a character is missing.

**Reply by lojjic** [reply]
Oh, I didn't realize you don't have control over the fonts being used -- so your users can both enter arbitrary text _and_ use arbitrary fonts. Got it.

Digging deeper into your requirements ... would you need to know _which_ characters are missing a glyph mapping? Or just a single yes/no whether there are missing glyphs or not?

If you're game for working up a PR ... the [`glyphIds` array here](https://github.com/protectwise/troika/blob/a6b913e475a3a8459b2f758c709b8feb16966ed1/packages/troika-three-text/src/TextBuilder.js#L179) is probably a good source -- if that array contains `0` for any glyph that indicates a missing character. (Glyph index `0` is the "missing character glyph" according to the font specs.) So you could check for it in that loop, assuming a simple true/false, and pass it to the callback later on. Then it'll be available in your Font object's `textRenderInfo`.

**Reply by freakypie** [reply]
I might be able to help. Right now, i just compare the glyph mapping to the mapping of a kanji that is unlikely to be in the font set. It's a hack, but it works for now.

My app is a game and the text is important. So i create each character separately for animation. Ideal for me might not fit your standards due to that. I will take a shot at a PR though
